### ISSUE TITLE
xnn_params.f32.hwc2spchw_dconv3x3c3s2.ukernel_with_symm_padding is NULL

### ISSUE THREAD

**Issue by WeiHao97** [google/XNNPACK#574]
Hi!
I am using X86 desktop.
when I try to create the full convolution which takes nhwc as input and outputs nchw:
xnn_create_convolution2d_nchw_f32(
    1 /* top padding */, 1 /* right padding */,
    1 /* bottom padding */, 1 /* left padding */,
    3 /* kernel height */, 3 /* kernel width */,
    2 /* subsampling height */, 2 /* subsampling width */,
    1 /* dilation_height */, 1 /* dilation_width */,
    1 /* groups */,
    3 /* input channels per group */,
    24 /* output_channels_per_group */,
    w0, w1,
    0.0f /* output min */, 6.0f /* output max */,
    XNN_FLAG_INPUT_NHWC/* flags */,
    &op0);

en error occured: failed to create Convolution operator: only selected Convolution parameters are supported
and I found that it is because xnn_params.f32.hwc2spchw_dconv3x3c3s2.ukernel_with_symm_padding == NULL which should be initialized for x86

I tried to exclude 'XNN_NO_NCHW_OPERATORS' in BUILD.bazel for the "xnnpack_operators_nhwc_f32" library but received the same error.

**Comment by Maratyszcza** [comment]
This function is implemented only for ARM64 and WebAssembly

**Comment by WeiHao97** [comment]
Hi,
I switched to a bare-metal ARM64 (X-GENE, aarch64) machine running Ubuntu 16.04 (xenial). However,  an error ”error: unknown value ‘armv8.2-a+fp16’ for -march“ occured when I run 'make end2end-bench'.
I checked issue[ #323](https://github.com/google/XNNPACK/issues/323) and [#242](https://github.com/google/XNNPACK/issues/242) but still not sure how to solve this.

**Comment by Maratyszcza** [comment]
You need a newer compiler (e.g. gcc 7+). Older versions don’t support FP16 arithmetic instructions in some of XNNPACK microkernels.

**Comment by WeiHao97** [comment]
Hi @Maratyszcza ,
It seems I made the sparse version work. However, I have a monor question about the paddings in some layers in the mobilenet where they are using 0 /* top padding */, 1 /* right padding */, 1 /* bottom padding */, 0 /* left padding */. In your current nchw implementation,  left and right paddings are required to be all 1's which cannot be satisfied for the aformationed layers. Though it seems running with no problem even if I creat the layer with  1 /* top padding */, 1 /* right padding */, 0 /* bottom padding */, 1 /* left padding */, the output will be a little different. I am wondering if your team is considering making more padding patterns in the implementation. Thanks!

**Comment by Maratyszcza** [comment]
MobileNet in TFLite model zoo was trained with TensorFlow SAME padding, which results in asymmetric padding for convolutions with non-unit stride. To get padding 1 one all sides in TF you need to explicitly pad the convolution input and set convolution padding mode to VALID.

**Comment by WeiHao97** [comment]
Hi @Maratyszcza,
Sorry that I did not make my question clear. I am building the mobilenet v1 with sparse weight only using the XNNPACK library. I am using the template in the "models" folder and replacing the nhwc operators with the nchw's which can handle the sparse weights. In the mobile net v1, there is a layer looks like:
  xnn_operator_t op3 = nullptr;
  status = xnn_create_convolution2d_nhwc_f32(
    0 /* top padding */, 1 /* right padding */,
    1 /* bottom padding */, 0 /* left padding */,
    3 /* kernel height */, 3 /* kernel width */,
    2 /* subsampling height */, 2 /* subsampling width */,
    1 /* dilation_height */, 1 /* dilation_width */,
    48 /* groups */,
    1 /* input channels per group */,
    1 /* output_channels_per_group */,
    48 /* input pixel stride */,
    48 /* output pixel stride */,
    w6, w7,
    0.0f /* output min */, 6.0f /* output max */,
    0 /* flags */,
    &op3);

which requests  asymmetric padding that is not implemented in the nchw's operators.  it seems running with no problem even if I creat the layer with 1 / top padding /, 1 / right padding /, 0 / bottom padding /, 1 / left padding */, but I think the output will be a little different. I am wondering if your team is considering making more padding patterns in the implementation. Thanks!

**Comment by Maratyszcza** [comment]
@WeiHao97 Are you calling `xnn_create_convolution2d_ncwc_f32`? It allows top padding of 0, but all other paddings must be 1, see https://github.com/google/XNNPACK/blob/1f29b806ebd61c56d9ad9b4dd36ea387d0c46cc7/src/operators/convolution-nchw.c#L171-L172

**Comment by WeiHao97** [comment]
> @WeiHao97 Are you calling `xnn_create_convolution2d_ncwc_f32`? It allows top padding of 0, but all other paddings must be 1, see
> 
> 
Yes, My question is how should I replace this nhwc operator : howhttps://github.com/google/XNNPACK/blob/1f29b806ebd61c56d9ad9b4dd36ea387d0c46cc7/src/operators/convolution-nchw.c#L171-L172

Since it is using asymmetric padding

**Comment by Maratyszcza** [comment]
You need a convolution operator with explicit padding. TensorFlow doesn't provide such primitive, but you can simulate it via a combination of `tf.pad` and a convolution with VALID padding. You may look at our [pre-trained models](https://github.com/google-research/google-research/tree/master/fastconvnets) for an example.

**Comment by WeiHao97** [comment]
> You need a convolution operator with explicit padding. TensorFlow doesn't provide such primitive, but you can simulate it via a combination of `tf.pad` and a convolution with VALID padding. You may look at our [pre-trained models](https://github.com/google-research/google-research/tree/master/fastconvnets) for an example.

Sorry, I am confused. Do you mean that it is not possible to build the mobilenet v1 handling sparse weight only using the XNNPACK library (as the way it is done with the [nhwc version](https://github.com/google/XNNPACK/blob/master/models/mobilenet-v1.cc) )?

**Comment by Maratyszcza** [comment]
It is possible to build MobileNet v1 in XNNPACK. But XNNPACK operators don't necessarily map 1:1 to TensorFlow or TFLite operators. In particular, XNNPACK operator with explicit padding generally maps to two operators in TF/TFLite (PAD + CONV_2D).

**Comment by WeiHao97** [comment]
> It is possible to build MobileNet v1 in XNNPACK. But XNNPACK operators don't necessarily map 1:1 to TensorFlow or TFLite operators. In particular, XNNPACK operator with explicit padding generally maps to two operators in TF/TFLite (PAD + CONV_2D).

So, in case of building MobileNet v1 in XNNPACK that handles sparse weight, my understanding is to replace every operator in the [mobilenet-v1.cc](https://github.com/google/XNNPACK/blob/master/models/mobilenet-v1.cc) with the corresponding operator that takes nchw format input. However, [some layers](https://github.com/google/XNNPACK/blob/1f29b806ebd61c56d9ad9b4dd36ea387d0c46cc7/models/mobilenet-v1.cc#L237) requests asymmetric padding that is not allowed in the [nchw setting](https://github.com/google/XNNPACK/blob/1f29b806ebd61c56d9ad9b4dd36ea387d0c46cc7/src/operators/convolution-nchw.c#L171-L172). Do you have any suggestions on how should I deal with this problem? Thanks!

**Comment by Maratyszcza** [comment]
When you train a MobileNet model, emulate convolution with explicit padding via a combination of pad operator (to pad input image with 1 pixel on each side) and convolution operator with VALID padding. When you convert the model to run on XNNPACK, replace the pad+convolution with XNNPACK Convolution operator with padding_left = padding_right = padding_top = padding_bottom = 1.

**Comment by Maratyszcza** [comment]
Sparse inference is now integrated in the TensorFlow Lite XNNPACK backend, see https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/delegates/xnnpack/README.md#sparse-inference-experimental for instructions.

**Comment by WeiHao97** [comment]
Thank you!
